### PR TITLE
Protect environment variables

### PR DESCRIPTION
**Pull request by adrien-cotte**
Hi,

Here is my PR for issue https://github.com/cea-hpc/modules/issues/429.
Let me know if something is wrong.

Best,
Adrien

### CODE REVIEW

File: tcl/envmngt.tcl.in [hunk 1]
```diff
@@ -1166,27 +1166,42 @@ proc get-env {var {valifunset {}}} {
    }
 }
 
+proc isProtectedEnvVar {var} {
+   # return if var is protected by protected_envvars config
+   foreach protectedvar [split [getConf protected_envvars] :] {
+      if {$var eq $protectedvar} {
```
Comment (xdelaruelle) [inline tcl/envmngt.tcl.in:1172]: should use `$var in [split [getConf protected_envvars] :]` rather foreach+if

File: tcl/envmngt.tcl.in [hunk 2]
```diff
@@ -1166,27 +1166,42 @@ proc get-env {var {valifunset {}}} {
...
+         reportDebug "$var not set (protected_envvars config)"
+         return true
+      }
+   }
+   return false
```
Comment (xdelaruelle) [inline tcl/envmngt.tcl.in:1177]: should simply return the result obtained from test:
`return [expr {$var in [split [getConf protected_envvars] :]}]`

File: tcl/envmngt.tcl.in [hunk 3]
```diff
@@ -1202,22 +1217,26 @@ proc unset-env {var {internal 0} {val {}}} {
...
-      interp-sync-env unset $var
-      set intwasset 1
-   }
+   if {![isProtectedEnvVar $var]} {
+      # clear value instead of unset it not to break variable later reference
+      # in modulefile. clear whether variable set or not to get a later usage
+      # consistent behavior whatever env is setup
+      if {!$internal} {
+         reset-to-unset-env $var $val
+      # internal variables (like ref counter var) are purely unset if they exists
```
Comment (xdelaruelle) [inline tcl/envmngt.tcl.in:1226]: line should not exceed 78 col length

File: tcl/envmngt.tcl.in [hunk 4]
```diff
@@ -1166,27 +1166,42 @@ proc get-env {var {valifunset {}}} {
...
-      }
+         # variable is not cleared anymore if set again
+         if {[info exists ::g_clearedEnvVars($var)]} {
+            unset ::g_clearedEnvVars($var)
+         }
 
-      # propagate variable setup to shell environment on load and unload mode
-      if {$mode eq {load} || $mode eq {unload}} {
-         set ::g_stateEnvVars($var) new
+         # propagate variable setup to shell environment on load and unload mode
```
Comment (xdelaruelle) [inline tcl/envmngt.tcl.in:1198]: line should not exceed 78 col length

**Review by xdelaruelle — CHANGES_REQUESTED** [review]
Thanks for the PR.

In addition to the review comments, new tests should be added to cover code changes as described in https://modules.readthedocs.io/en/latest/design/add-new-config-option.html#testsuite

File: tcl/envmngt.tcl.in (hunk 2 above)
Comment (adrien-cotte) [reply tcl/envmngt.tcl.in:1162]: Okay, but what about `reportDebug` call?
Should I shift it to `set-env` and `unset-env` statements?

Comment (xdelaruelle) [reply tcl/envmngt.tcl.in:1162]: no need, you already have the reportWarning there

**Comment by xdelaruelle** [comment]
This is progressing well. I have changed a little bit the doc and the warning message.

It is now time to craft specific tests to validate the correct behavior of the configuration when not set, set with valid or invalid values. Tests should cover attempts to modify variables with `append-path`, `prepend-path`, `remove-path`, `setenv`, `unsetenv` and `source-sh`.

Tests should cover when these changes occur on the different modulefile evaluation mode (load, unload, help, display, test, refresh, whatis). Test should also cover when protected variables are modified by `append-path`, `prepend-path` and `remove-path` sub-commands.

Also there should be a test when Modules-specific variable (like `LOADEDMODULES`) is defined in `protected_envvars`.

**Comment by adrien-cotte** [comment]
Hey @xdelaruelle, many thanks for your assistance!

> It is now time to craft specific tests to validate the correct behavior of the configuration when not set, set with valid or invalid values. Tests should cover attempts to modify variables with `append-path`, `prepend-path`, `remove-path`, `setenv`, `unsetenv` and `source-sh`.
> 
> Tests should cover when these changes occur on the different modulefile evaluation mode (load, unload, help, display, test, refresh, whatis). Test should also cover when protected variables are modified by `append-path`, `prepend-path` and `remove-path` sub-commands.

Agreed, but could you help me (again) to find the right files to modify or create, please?

> Also there should be a test when Modules-specific variable (like `LOADEDMODULES`) is defined in `protected_envvars`.

Good point, which behavior do you want for those use cases?
Maybe we should return an error if `protected_envvars` contains a "forbidden var" (`LOADEDMODULES`, `MODULEPATH`, `MODULESHOME`, `MODULES_CMD`, `_LMFILES_` etc...).
Is there a easy way to get an exhaustive list?

**Comment by xdelaruelle** [comment]
> Good point, which behavior do you want for those use cases?
> Maybe we should return an error if `protected_envvars` contains a "forbidden var" (`LOADEDMODULES`, `MODULEPATH`, `MODULESHOME`, `MODULES_CMD`, `_LMFILES_` etc...).
> Is there a easy way to get an exhaustive list?

For the initial implementation maybe it could be left as is, so if a user adds a Modules-specific variable in the protected list, module will not work correctly anymore. Will see if such strategy will fit on the long term. We could look later on at implementing a special treatment for these variables.

**Comment by xdelaruelle** [comment]
> > It is now time to craft specific tests to validate the correct behavior of the configuration when not set, set with valid or invalid values. Tests should cover attempts to modify variables with `append-path`, `prepend-path`, `remove-path`, `setenv`, `unsetenv` and `source-sh`.
> > Tests should cover when these changes occur on the different modulefile evaluation mode (load, unload, help, display, test, refresh, whatis). Test should also cover when protected variables are modified by `append-path`, `prepend-path` and `remove-path` sub-commands.
> 
> Agreed, but could you help me (again) to find the right files to modify or create, please?

I suggest you create a new test file in `modules.70-maint`, named `410-protected_envvars.exp`. You could base its content on `modules.50-cmds/540-complete.exp`.

Along this test file, I suggest you add a `protect/1.0` modulefile in `modulefiles.3` modulepath. Content of this modulefile could be based on `complete/1.0`. Having a `TESTSUITE_PROTECT` selection variable instead of `TESTSUITE_COMPLETE`. Each case of the switch in this modulefile may corresponds to a modulefile command to test (append-path, setenv, unsetenv, etc).

**Comment by xdelaruelle** [comment]
@adrien-cotte I have just updated your branch to update the documentation commit and make it comply with the content that is now expected for configuration option and environment variable description.

**Comment by xdelaruelle** [comment]
@adrien-cotte Sorry, I have made a mistake when trying to re-update your branch against the `master` branch update I have pushed this morning. I have mistakenly pushed the current master branch against your master branch, which was also used for this PR. It has closed the PR. I have re-uploaded your branch on [my test repository](https://github.com/xdelaruelle/modules/tree/protect), so you can fetch there to include it in your github repo and recreate a PR from that.

**Pull request closed without merging**